usr/sbin/p11sak/p11sak.c

```diff
@@ -3481,162 +3481,321 @@ static const struct p11sak_objtype *find_certtype(CK_KEY_TYPE ktype)
-        { CKA_CLASS, &class_val, sizeof(class_val) },
-        { CKA_KEY_TYPE, &ktype_val, sizeof(ktype_val) },
-    };
-    const CK_ULONG num_attrs = sizeof(attrs) / sizeof(CK_ATTRIBUTE);
-    const struct p11sak_objtype *keytype_val;
-    CK_ULONG keysize_val = 0;
-    CK_ATTRIBUTE keysize_attr;
-    int rv;
 
-    rc = pkcs11_funcs->C_GetAttributeValue(pkcs11_session, key,
-                                           attrs, num_attrs);
-    if (rc != CKR_OK && rc != CKR_ATTRIBUTE_TYPE_INVALID &&
-        rc != CKR_ATTRIBUTE_SENSITIVE) {
-        warnx("Failed to get attributes: C_GetAttributeValue: 0x%lX: %s",
-              rc, p11_get_ckr(rc));
+    rc = get_attribute(obj, &attr);
+    if (rc != CKR_OK) {
+        warnx("Failed to retrieve attribute CKA_VALUE from object "
+              "\"%s\": 0x%lX: %s", label, rc, p11_get_ckr(rc));
         return rc;
     }
 
-    if (attrs[0].ulValueLen != 0 &&
-        attrs[0].ulValueLen != CK_UNAVAILABLE_INFORMATION) {
-        attrs[0].pValue = calloc(attrs[0].ulValueLen + 1, 1);
-        if (attrs[0].pValue == NULL) {
-            warnx("Failed to allocate memory for label attribute");
+    if (attr.ulValueLen == 0 || attr.ulValueLen == CK_UNAVAILABLE_INFORMATION) {
+        warnx("CKA_VALUE of object \"%s\" is unavailable: 0x%lX: %s",
+              label, rc, p11_get_ckr(rc));
+        *common_name_value = strdup("");
+        if (*common_name_value == NULL) {
+            warnx("Failed to allocate memory for common_name_value "
+                  "for object \"%s\"", label);
             return CKR_HOST_MEMORY;
         }
+        return CKR_OK;
+    }
 
-        rc = pkcs11_funcs->C_GetAttributeValue(pkcs11_session, key,
-                                               attrs, num_attrs);
-        if (rc != CKR_OK && rc != CKR_ATTRIBUTE_TYPE_INVALID &&
-            rc != CKR_ATTRIBUTE_SENSITIVE) {
-            warnx("Failed to get attributes: C_GetAttributeValue: 0x%lX: %s",
-                  rc, p11_get_ckr(rc));
-            free(attrs[0].pValue);
-            return rc;
-        }
-    } else {
-        attrs[0].pValue = strdup("");
-        if (attrs[0].pValue == NULL) {
-            warnx("Failed to allocate memory for label attribute");
-            return CKR_HOST_MEMORY;
+    tmp_ptr = attr.pValue;
+    x509 = d2i_X509(NULL, &tmp_ptr, attr.ulValueLen);
+    if (x509 == NULL) {
+        *common_name_value = strdup("");
+        if (*common_name_value == NULL) {
+            warnx("Failed to allocate memory for common_name_value "
+                  "for object \"%s\"", label);
+            rc = CKR_HOST_MEMORY;
+        } else {
+            rc = CKR_OK;
         }
+        goto done;
     }
 
-    switch (class_val) {
-    case CKO_PUBLIC_KEY:
-    case CKO_PRIVATE_KEY:
-    case CKO_SECRET_KEY:
-        break;
-    default:
-        free(attrs[0].pValue);
-        return CKR_KEY_NEEDED;
+    subj = X509_NAME_oneline(X509_get_subject_name(x509), NULL, 0);
```

Comment:
What if X509_get_subject_name() returns NULL ?

Reply:
I had the same thought, but the function just returns an internal pointer, so I'm assuming it cannot be NULL. Same with X509_NAME_oneline. 
https://www.openssl.org/docs/man1.1.1/man3/X509_get_subject_name.html

Reply:
Done, as above.